### PR TITLE
script_defs: open state files and deployments with a context manager

### PR DESCRIPTION
**Pull request by grahamc**
Use context managers to open and manage statefiles and deployments. This lets us do some fancier things with those state files. For example, storing the state other places.

This change is almost entirely whitespace. Check it out without whitespace in the diff: https://github.com/NixOS/nixops/pull/1263/files?w=1

Also consider looking commit-by-commit, which makes it easier, I think, to review.

Here is a diff adding state backends using this: legacy, in-memory, and s3: https://github.com/grahamc/nixops/compare/context-manager...grahamc:context-manager-states-rebase?expand=1

### CODE REVIEW

File: nixops/script_defs.py [hunk 1]
```diff
@@ -782,157 +818,166 @@ def scp_loc(user, ssh_name, remote, loc):
...
 
 
 def op_list_generations(args):
-    depl = check_rollback_enabled(args)
-    if (
-        subprocess.call(["nix-env", "-p", depl.get_profile(), "--list-generations"])
-        != 0
-    ):
-        raise Exception("nix-env --list-generations failed")
+    with with_rollback_enabled(args) as depl:
```
Comment (adisbladis) [inline nixops/script_defs.py]: Maybe this is nitpicking but I feel that `with with_` is a bit strange.

Comment (grahamc) [reply nixops/script_defs.py]: Oh yeah, totally. I don't really like the names of these context managers. Maybe you have a different suggestion?

Comment (adisbladis) [reply nixops/script_defs.py]: Maybe invert the naming and call it `deployment_with_rollback`? It also makes things more discoverable.

Comment (grahamc) [reply nixops/script_defs.py]: Fixed!

File: nixops/script_defs.py [hunk 2]
```diff
@@ -37,6 +37,22 @@
...
+        depl = open_deployment(sf, args)
+        yield depl
+
+
+@contextlib.contextmanager
+def network_state(args: Namespace) -> Generator[nixops.statefile.StateFile, None, None]:
+    state = nixops.statefile.StateFile(args.state_file)
+    try:
+        yield state
+    finally:
```
Comment (grahamc) [inline nixops/script_defs.py]: This used to catch the exception and no longer tries to.

File: nixops/script_defs.py [hunk 3]
```diff
@@ -782,157 +815,166 @@ def scp_loc(user, ssh_name, remote, loc):
...
+        env["NIX_SSHOPTS"] = " ".join(m.get_ssh_flags())
+        res = nixops.util.logged_exec(
+            [
+                "nix",
+                "copy",
+                "--to",
+                "ssh://{}".format(m.get_ssh_name()),
+                args.storepath,
+            ],
+            m.logger,
```
Comment (domenkozar) [inline nixops/script_defs.py]: :+1:

**Pull request merged**